src/track/bpm.h

```diff
@@ -83,25 +83,87 @@ class Bpm final {
         return displayValueText(m_value);
     }
 
+    Bpm& operator+=(double increment) {
+        m_value += increment;
+        return *this;
+    }
+
+    Bpm& operator-=(double decrement) {
+        m_value -= decrement;
+        return *this;
+    }
+
+    Bpm& operator*=(double multiple) {
+        m_value *= multiple;
+        return *this;
+    }
+
+    Bpm& operator/=(double divisor) {
+        m_value /= divisor;
+        return *this;
+    }
+
 private:
     double m_value;
 };
 
-inline
-bool operator==(const Bpm& lhs, const Bpm& rhs) {
-    return lhs.compareEq(rhs);
+/// Bpm can be added to a double
+inline Bpm operator+(Bpm bpm, double bpmDiff) {
+    return Bpm(bpm.getValue() + bpmDiff);
+}
+
+/// Bpm can be subtracted from a double
+inline Bpm operator-(Bpm bpm, double bpmDiff) {
+    return Bpm(bpm.getValue() - bpmDiff);
 }
 
-inline
-bool operator!=(const Bpm& lhs, const Bpm& rhs) {
-    return !(lhs == rhs);
+/// Two Bpm values can be subtracted to get a double
+inline double operator-(Bpm bpm1, Bpm bpm2) {
+    return bpm1.getValue() - bpm2.getValue();
 }
 
-inline
-QDebug operator<<(QDebug dbg, const Bpm& arg) {
-    return dbg << arg.getValue();
+// Adding two Bpm is not allowed, because it makes no sense semantically.
+
+/// Bpm can be multiplied or divided by a double
+inline Bpm operator*(Bpm bpm, double multiple) {
+    return Bpm(bpm.getValue() * multiple);
 }
 
+inline Bpm operator/(Bpm bpm, double divisor) {
+    return Bpm(bpm.getValue() / divisor);
+}
+
+inline bool operator<(Bpm bpm1, Bpm bpm2) {
+    return bpm1.getValue() < bpm2.getValue();
+}
+
+inline bool operator<=(Bpm bpm1, Bpm bpm2) {
+    return bpm1.getValue() <= bpm2.getValue();
+}
+
+inline bool operator>(Bpm bpm1, Bpm bpm2) {
+    return bpm1.getValue() > bpm2.getValue();
+}
+
+inline bool operator>=(Bpm bpm1, Bpm bpm2) {
+    return bpm1.getValue() >= bpm2.getValue();
+}
+
+inline bool operator==(Bpm bpm1, Bpm bpm2) {
```

Comment:
I didn't notice this very special quality comparison! We should add a comment that all invalid values are considered equal.

We also should test the comparison of invalid and valid values.

Reply:
I will. But do you agree with this comparison? I think it makes sense to consider all invalid values equal.

Reply:
I'm not saying we should imitate this...
```
❯ node
Welcome to Node.js v14.17.0.
Type ".help" for more information.
> NaN === NaN
false
>
```
https://www.destroyallsoftware.com/talks/wat

Reply:
This is actually the IEEE-754 standard. It's the same in C++. `x == x` is another way to implement `isNaN`.

Reply:
It could be surprising as the other comparison operators like `<=` behave differently! Do we really need it? Otherwise I would suggest to avoid these tweaks.

Reply:
@uklotzde test added.

Reply:
> It could be surprising as the other comparison operators like `<=` behave differently! Do we really need it? Otherwise I would suggest to avoid these tweaks.

I don't know. My rationale was that `someValue == Bpm()` is true if `someValue` is some invalid value.

Btw, I actually have a WIP PR that takes care of that and raises a debug assertion every time you try to do a calculation or comparison with an invalid value: https://github.com/mixxxdj/mixxx/pull/4055/commits/221747015dc1dc8b9f85a99e6d9f4864a5416a6f#diff-6ee8fe6454f9e7e8637c1ba82b07d85ae618df4fbb651d66f8e5fff4b06f51e5

Reply:
> It could be surprising as the other comparison operators like `<=` behave differently! Do we really need it? Otherwise I would suggest to avoid these tweaks.

I don't know. My rationale was that `someValue == Bpm()` is true if `someValue` is some invalid value.

I actually have a WIP PR that takes care of that and raises a debug assertion every time you try to do a calculation or comparison with an invalid value: https://github.com/mixxxdj/mixxx/pull/4055/commits/221747015dc1dc8b9f85a99e6d9f4864a5416a6f#diff-6ee8fe6454f9e7e8637c1ba82b07d85ae618df4fbb651d66f8e5fff4b06f51e5

But I don't want to add all of that here. I can also revert these "all invalid values are equal" changes if you prefer.